### PR TITLE
[MIRROR] MOD update: Modular Cores [MDB IGNORE]

### PR DESCRIPTION
**Pull request by SkyratBot**
Original PR: https://github.com/tgstation/tgstation/pull/64042
--------------------

## About The Pull Request

minor changes:
- improves the springlock module by spreading damage across limbs, but making the equipping buff it gives better
- improves the rave module by using color matrices, and fixes it crashing the tgui
- fixes teleport module fucking up your matrices
- adds advanced storage modules to the techweb
- fixes not being able to spraypaint your suit (or any storage item for that matter)
- fixes berserker suit having infinite charge for real
- makes armor booster not remove your heat protection (rip elite nukies)
- balances bomb armor on some suits
- makes it so some specific suits protect their contents from explosions (research, syndicate, captain, admin)
- fixes #63945
- fixes #63973

bigger changes:
1. The suit slowdown is now 1/5 per piece, including the control unit, instead of 1/4 per piece, so undeployed suits will have a minor slowdown
2. You can now right click the deploy and activate action buttons. Right clicking activation will skip the prompt if you want to use it. Right clicking deploy will try to deploy/retract all pieces at once.
3. Adds support for modular cores. They can power your suit in a different way and in the future add some other cool stuff. Currently there are 3 core types:
- Standard: These use ethereal hearts to convert cell charge into viable sources of power. Your current experience.
- Infinite: Infinite charge palladium core. Used by some admin suits and the wizard suit.
- Ethereal: Use power directly from ethereal bodies, not requiring any power cell at all, only work on beings with ethereal stomachs, craftable.

image

https://streamable.com/0p3qa5

## Why It's Good For The Game

makes the suits more modular, adds support for more unique suits like blood powered cult stuff, adds a quick way to use them, makes them not a better backpack by giving them at least a bit of a slowdown

## Changelog

:cl: Fikou
qol: You can now right-click modsuit action buttons to activate them faster.
balance: Undeployed modsuits now have 1/5 of the slowdown instead of 0 (from 0.15 on quick suits like rescue and 0.2 on semi quick like advanced or medical to 0.35 on research and 0.4 on prototype)
qol: the storage modsuit module now tries to scoop up the suit storage item
balance: you can now print expanded storage modules
fix: the magnate suit now has a jetpack
fix: you can now spraypaint storage items
fix: fixes bomb protecting modsuits not protecting the contents from bombs
fix: fixes elite suit nukies burning down
add: MODsuit modular cores. These power your suit, the three types of them are: standard, your current ones, using power cells. infinite, dont use power at all, used by some admin and wizard suits. ethereal, these dont use power cells, but an ethereal's charge directly, they are craftable.
/:cl:

### CODE REVIEW

File: code/__DEFINES/dcs/signals/signals_mod.dm [hunk 1]
```diff
@@ -9,3 +9,16 @@
...
+#define COMSIG_MODULE_TRIGGERED "mod_module_triggered"
+	// Cancels activation, with no message. include feedback on your cancel.
+	#define MOD_ABORT_USE (1<<0)
+/// Called when a module activates, after all checks have passed and cooldown started.
+#define COMSIG_MODULE_ACTIVATED "mod_module_activated"
+/// Called when a module deactivates, after all checks have passed.
+#define COMSIG_MODULE_DEACTIVATED "mod_module_deactivated"
+/// Called when a module is used, after all checks have passed and cooldown started.
+#define COMSIG_MODULE_USED "mod_module_used"
+>>>>>>> fd9a7f8a581 (MOD update: Modular Cores (#64042))
```
Comment (OrionTheFox) [inline code/__DEFINES/dcs/signals/signals_mod.dm:24]: I honestly dont see the "Conflict" here, it just... didnt add these to the file right?

```suggestion

/// Called when a module attempts to activate, however it does. At the end of checks so you can add some yourself, or work on trigger behavior (mob/user)
#define COMSIG_MODULE_TRIGGERED "mod_module_triggered"
	// Cancels activation, with no message. include feedback on your cancel.
	#define MOD_ABORT_USE (1<<0)
/// Called when a module activates, after all checks have passed and cooldown started.
#define COMSIG_MODULE_ACTIVATED "mod_module_activated"
/// Called when a module deactivates, after all checks have passed.
#define COMSIG_MODULE_DEACTIVATED "mod_module_deactivated"
/// Called when a module is used, after all checks have passed and cooldown started.
#define COMSIG_MODULE_USED "mod_module_used"

```

File: code/modules/clothing/outfits/standard.dm [hunk 2]
```diff
@@ -414,6 +414,16 @@
...
+<<<<<<< HEAD
+=======
+	back = /obj/item/mod/control/pre_equipped/chrono
+
+/datum/outfit/chrono_agent/post_equip(mob/living/carbon/human/agent, visualsOnly)
+	. = ..()
+	var/obj/item/mod/control/mod = agent.back
+	var/obj/item/mod/module/eradication_lock/lock = locate(/obj/item/mod/module/eradication_lock) in mod.modules
+	lock.true_owner_ckey = agent.ckey
+>>>>>>> fd9a7f8a581 (MOD update: Modular Cores (#64042))
```
Comment (OrionTheFox) [inline code/modules/clothing/outfits/standard.dm:426]: merge this PR after https://github.com/Skyrat-SS13/Skyrat-tg/pull/10682 apparently

```suggestion
	back = /obj/item/mod/control/pre_equipped/chrono

/datum/outfit/chrono_agent/post_equip(mob/living/carbon/human/agent, visualsOnly)
	. = ..()
	var/obj/item/mod/control/mod = agent.back
	var/obj/item/mod/module/eradication_lock/lock = locate(/obj/item/mod/module/eradication_lock) in mod.modules
	lock.true_owner_ckey = agent.ckey
```

File: code/modules/mod/mod_clothes.dm [hunk 3]
```diff
@@ -3,24 +3,18 @@
 	desc = "A helmet for a MODsuit."
 	icon = 'icons/obj/mod.dmi'
 	icon_state = "helmet"
+<<<<<<< HEAD
 	worn_icon = 'icons/mob/mod.dmi'
 	armor = list(MELEE = 0, BULLET = 0, LASER = 0, ENERGY = 0, BOMB = 0, BIO = 100, FIRE = 25, ACID = 25, WOUND = 10)
+=======
+	worn_icon = 'icons/mob/clothing/mod.dmi'
+	armor = list(MELEE = 0, BULLET = 0, LASER = 0, ENERGY = 0, BOMB = 0, BIO = 0, FIRE = 0, ACID = 0, WOUND = 0)
+>>>>>>> fd9a7f8a581 (MOD update: Modular Cores (#64042))
```
Comment (OrionTheFox) [inline code/modules/mod/mod_clothes.dm:12]: icon path change & armor value change on the unused basetype

```suggestion
	worn_icon = 'icons/mob/clothing/mod.dmi'
	armor = list(MELEE = 0, BULLET = 0, LASER = 0, ENERGY = 0, BOMB = 0, BIO = 0, FIRE = 0, ACID = 0, WOUND = 0)
```

File: code/modules/mod/mod_clothes.dm [hunk 4]
```diff
@@ -70,16 +57,17 @@
 	desc = "A pair of gauntlets for a MODsuit."
 	icon = 'icons/obj/mod.dmi'
 	icon_state = "gauntlets"
+<<<<<<< HEAD
 	worn_icon = 'icons/mob/mod.dmi'
 	armor = list(MELEE = 0, BULLET = 0, LASER = 0, ENERGY = 0, BOMB = 0, BIO = 100, FIRE = 25, ACID = 25, WOUND = 10)
+=======
+	worn_icon = 'icons/mob/clothing/mod.dmi'
+	armor = list(MELEE = 0, BULLET = 0, LASER = 0, ENERGY = 0, BOMB = 0, BIO = 0, FIRE = 0, ACID = 0, WOUND = 0)
+>>>>>>> fd9a7f8a581 (MOD update: Modular Cores (#64042))
```
Comment (OrionTheFox) [inline code/modules/mod/mod_clothes.dm:66]: icon path change & armor value change on the unused basetype

```suggestion
	worn_icon = 'icons/mob/clothing/mod.dmi'
	armor = list(MELEE = 0, BULLET = 0, LASER = 0, ENERGY = 0, BOMB = 0, BIO = 0, FIRE = 0, ACID = 0, WOUND = 0)
```

File: code/modules/mod/mod_clothes.dm [hunk 5]
```diff
@@ -108,16 +96,19 @@
 	desc = "A pair of boots for a MODsuit."
 	icon = 'icons/obj/mod.dmi'
 	icon_state = "boots"
+<<<<<<< HEAD
 	worn_icon = 'icons/mob/mod.dmi'
 	armor = list(MELEE = 0, BULLET = 0, LASER = 0, ENERGY = 0, BOMB = 0, BIO = 100, FIRE = 25, ACID = 25, WOUND = 10)
+=======
+	worn_icon = 'icons/mob/clothing/mod.dmi'
+	armor = list(MELEE = 0, BULLET = 0, LASER = 0, ENERGY = 0, BOMB = 0, BIO = 0, FIRE = 0, ACID = 0, WOUND = 0)
+>>>>>>> fd9a7f8a581 (MOD update: Modular Cores (#64042))
```
Comment (OrionTheFox) [inline code/modules/mod/mod_clothes.dm:105]: icon path change & armor value change (which is re-changed later BASED on the modsuit)

```suggestion
	worn_icon = 'icons/mob/clothing/mod.dmi'
	armor = list(MELEE = 0, BULLET = 0, LASER = 0, ENERGY = 0, BOMB = 0, BIO = 0, FIRE = 0, ACID = 0, WOUND = 0)
```

**Comment by ghost** [comment]
Im dumby and reviewed this too early;

Merge #10682 first and most of the conflicts should go away

**Pull request merged**